<html>
<head>
  <title>VEP summary</title>
  <script type="text/javascript" src="http://www.google.com/jsapi"></script>
  <script type="text/javascript">
    google.load('visualization', '1', {packages: ['corechart','table']});
  </script>
  <script type="text/javascript">
    
    function init() {
      // charts
      var var_class_pie = drawPie('var_class_pie', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['deletion',3],['insertionBND',10],['BND',11]]), null);
var var_class_table = drawTable('var_class_table', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['deletion',3],['insertionBND',10],['BND',11]]));

          google.visualization.events.addListener(var_class_pie, 'select', function() {
            var_class_table.setSelection(var_class_pie.getSelection());
          });
          google.visualization.events.addListener(var_class_table, 'select', function() {
            var_class_pie.setSelection(var_class_table.getSelection());
          });
        var var_cons_pie = drawPie('var_cons_pie', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6],['3_prime_UTR_variant',4],['non_coding_transcript_exon_variant',2],['intron_variant',2],['upstream_gene_variant',8],['downstream_gene_variant',1]]), {slices: [{color: "#ff69b4"}, {color: "#458b00"}, {}, {}, {color: "#02599c"}, {color: "#a2b5cd"}, {color: "#a2b5cd"}]});
var var_cons_table = drawTable('var_cons_table', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6],['3_prime_UTR_variant',4],['non_coding_transcript_exon_variant',2],['intron_variant',2],['upstream_gene_variant',8],['downstream_gene_variant',1]]));

          google.visualization.events.addListener(var_cons_pie, 'select', function() {
            var_cons_table.setSelection(var_cons_pie.getSelection());
          });
          google.visualization.events.addListener(var_cons_table, 'select', function() {
            var_cons_pie.setSelection(var_cons_table.getSelection());
          });
        var consequences_pie = drawPie('consequences_pie', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6],['3_prime_UTR_variant',4],['non_coding_transcript_exon_variant',2],['intron_variant',2],['upstream_gene_variant',37],['downstream_gene_variant',24],['feature_elongation',7],['feature_truncation',2]]), {slices: [{color: "#ff69b4"}, {color: "#458b00"}, {}, {}, {color: "#02599c"}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {}, {}]});
var consequences_table = drawTable('consequences_table', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6],['3_prime_UTR_variant',4],['non_coding_transcript_exon_variant',2],['intron_variant',2],['upstream_gene_variant',37],['downstream_gene_variant',24],['feature_elongation',7],['feature_truncation',2]]));

          google.visualization.events.addListener(consequences_pie, 'select', function() {
            consequences_table.setSelection(consequences_pie.getSelection());
          });
          google.visualization.events.addListener(consequences_table, 'select', function() {
            consequences_pie.setSelection(consequences_table.getSelection());
          });
        var coding_pie = drawPie('coding_pie', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6]]), {slices: [{color: "#ff69b4"}, {color: "#458b00"}]});
var coding_table = drawTable('coding_table', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['frameshift_variant',1],['coding_sequence_variant',6]]));

          google.visualization.events.addListener(coding_pie, 'select', function() {
            coding_table.setSelection(coding_pie.getSelection());
          });
          google.visualization.events.addListener(coding_table, 'select', function() {
            coding_pie.setSelection(coding_table.getSelection());
          });
        var chr_bar = drawBar('chr_bar', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['ChrA_C_glabrata_CBS138',16],['ChrB_C_glabrata_CBS138',8]]), {legend: {position: "none"}});
var chr_table = drawTable('chr_table', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['ChrA_C_glabrata_CBS138',16],['ChrB_C_glabrata_CBS138',8]]));

          google.visualization.events.addListener(chr_bar, 'select', function() {
            chr_table.setSelection(chr_bar.getSelection());
          });
          google.visualization.events.addListener(chr_table, 'select', function() {
            chr_bar.setSelection(chr_table.getSelection());
          });
        var chr_ChrA_C_glabrata_CBS138_area = drawArea('chr_ChrA_C_glabrata_CBS138_area', 'Distribution of variants on chromosome ChrA_C_glabrata_CBS138', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',16]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var chr_ChrB_C_glabrata_CBS138_area = drawArea('chr_ChrB_C_glabrata_CBS138_area', 'Distribution of variants on chromosome ChrB_C_glabrata_CBS138', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',8]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var protein_bar = drawBar('protein_bar', 'Position in protein', google.visualization.arrayToDataTable([['Position in protein (percentile)','Count'],['00-10%',0],['10-20%',0],['20-30%',0],['30-40%',0],['40-50%',0],['50-60%',2],['60-70%',0],['70-80%',1],['80-90%',2],['90-100%',2]]), {hAxis: {title: "Position in protein (percentile)", textStyle: {fontSize: 10}}, legend: {position: "none"}});

    }
    
    function drawPie(id, title, data, options) {    
      var pie = new google.visualization.PieChart(document.getElementById(id));
      pie.draw(data, options);
      return pie;
    }
    function drawBar(id, title, data, options) {
      var bar = new google.visualization.ColumnChart(document.getElementById(id));
      bar.draw(data, options);
      return bar;
    }
    function drawTable(id, title, data) {
      var table = new google.visualization.Table(document.getElementById(id));
      table.draw(data, null);
      return table;
    }
    function drawLine(id, title, data, options) {
      var line = new google.visualization.LineChart(document.getElementById(id));
      line.draw(data, options);
      return line;
    }
    function drawArea(id, title, data, options) {
      var area = new google.visualization.AreaChart(document.getElementById(id));
      area.draw(data, options);
      return area;
    }
    google.setOnLoadCallback(init);
  </script>
  
  
  <style type="text/css">
    body {
      font-family: arial, sans-serif;
      margin: 0px;
      padding: 0px;
    }
    
    a {color: #36b;}
    a.visited {color: #006;}
    
    .stats_table {
      margin: 5px;
    }
    
    td {
      padding: 5px;
    }

    th.gradient {
      height: auto;
    }

    .stats_table tr:nth-child(odd) {
      background-color: #f0f0f0;
    }
    
    h3 {
      color: #666;
    }
    
    .masthead {
      background-color: white;
      color: rgb(204, 221, 255);
      height: 80px;
      width: 100%;
      padding: 0px;
    }
    
    .main {
      padding: 10px;
    }
    
    .gradient {
      background: #333366; /* Old browsers */
      background: -moz-linear-gradient(left,  #333366 0%, #ffffff 100%); /* FF3.6+ */
      background: -webkit-gradient(linear, left top, right top, color-stop(0%,#333366), color-stop(100%,#ffffff)); /* Chrome,Safari4+ */
      background: -webkit-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Chrome10+,Safari5.1+ */
      background: -o-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Opera 11.10+ */
      background: -ms-linear-gradient(left,  #333366 0%,#ffffff 100%); /* IE10+ */
      background: linear-gradient(to right,  #333366 0%,#ffffff 100%); /* W3C */
      filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#333366', endColorstr='#ffffff',GradientType=1 ); /* IE6-9 */
      
      padding: 0px;
      height: 80px;
      width: 700px;
    }
    
    .main_content {
      margin-left: 300px;
    }
    
    .sidemenu {
      width: 260px;
      position: fixed;
      border-style: solid;
      border-width: 2px;
      border-color: rgb(51, 51, 102);
    }
    
    .sidemenu_head {
      width: 250px;
      background-color: rgb(51, 51, 102);
      color: rgb(204, 221, 255);
      padding: 5px;
    }
    
    .sidemenu_body {
      width: 250px;
      padding: 5px;
    }
  </style>
</head>
<body>
<div id="masthead" class="masthead">
  <div style="float: left; display: inline; padding: 10px; height: 80px;">
    <a href="http://www.ensembl.org/"><img src="http://static.ensembl.org/i/e-ensembl.png" width="185" height="45" alt="Ensembl"></a>
  </div>
  
  <div style="float: right; display: inline; height: 80px; background: white; padding: 10px;">
    <a href="http://www.ensembl.org/vep"><img src="http://www.ensembl.org/img/vep_logo.png" width="111" height="50" alt="VEP"></a>
  </div>
  <div class="gradient">
  </div>
</div>
<div class="main">
<div class="sidemenu"><div class="sidemenu_head">Links</div><div class="sidemenu_body"><ul><li><a href="#masthead">Top of page</a></li><li><a href="#run_stats">VEP run statistics</a></li><li><a href="#gen_stats">General statistics</a></li><li><a href="#var_class">Variant classes</a></li><li><a href="#var_cons">Consequences (most severe)</a></li><li><a href="#consequences">Consequences (all)</a></li><li><a href="#coding">Coding consequences</a></li><li><a href="#chr">Variants by chromosome</a></li><li><a href="#protein">Position in protein</a></li></ul></div></div><div class='main_content'><h3 id="run_stats">VEP run statistics</h3><table class="stats_table"><tr><td>VEP version (API)</td><td> 100 (100)</td></tr><tr><td>Annotation sources</td><td>Custom: /perSVade/installation/test_installation/testing_outputs/annotate_SVs/reference_genome_dir/reference_genome_features.gff_corrected.gff_with_biotype.gff_clean.gz (overlap)</td></tr><tr><td>Species</td><td>homo_sapiens</td></tr><tr><td>Command line options</td><td><pre>--custom /perSVade/installation/test_installation/testing_outputs/annotate_SVs/reference_genome_dir/reference_genome_features.gff_corrected.gff_with_biotype.gff_clean.gz,,gff --fasta /perSVade/installation/test_installation/testing_outputs/annotate_SVs/reference_genome_dir/reference_genome.fasta --fields Uploaded_variation,Location,Allele,Gene,Feature,Feature_type,Consequence,cDNA_position,CDS_position,Protein_position,Amino_acids,Codons,Existing_variation,Extra --force_overwrite --format vcf --gff /perSVade/installation/test_installation/testing_outputs/annotate_SVs/reference_genome_dir/reference_genome_features.gff_corrected.gff_with_biotype.gff_clean.gz --input_file /perSVade/installation/test_installation/testing_outputs/annotate_SVs/SV_and_CNV_variant_calling.vcf --output_file /perSVade/installation/test_installation/testing_outputs/annotate_SVs/SV_and_CNV_variant_calling.vcf_annotated_VEP.tab.tmp.raw.tbl.tmp --tab --verbose</pre></td></tr><tr><td>Start time</td><td>2022-07-04 11:00:39</td></tr><tr><td>End time</td><td>2022-07-04 11:00:39</td></tr><tr><td>Run time</td><td>0 seconds</td></tr><tr><td>Input file</td><td>/perSVade/installation/test_installation/testing_outputs/annotate_SVs/SV_and_CNV_variant_calling.vcf</td></tr><tr><td>Output file</td><td>/perSVade/installation/test_installation/testing_outputs/annotate_SVs/SV_and_CNV_variant_calling.vcf_annotated_VEP.tab.tmp.raw.tbl.tmp</td></tr></table><h3 id="gen_stats">General statistics</h3><table class="stats_table"><tr><td>Lines of input read</td><td>24</td></tr><tr><td>Variants processed</td><td>24</td></tr><tr><td>Variants filtered out</td><td>0</td></tr><tr><td>Novel / existing variants</td><td>-</td></tr><tr><td>Overlapped genes</td><td>12</td></tr><tr><td>Overlapped transcripts</td><td>12</td></tr><tr><td>Overlapped regulatory features</td><td>-</td></tr></table><hr/><h3 id="var_class">Variant classes</h3><div id="var_class_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="var_class_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="var_cons">Consequences (most severe)</h3><div id="var_cons_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="var_cons_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="consequences">Consequences (all)</h3><div id="consequences_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="consequences_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="coding">Coding consequences</h3><div id="coding_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="coding_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr">Variants by chromosome</h3><div id="chr_bar" style="width: 800px; height: 200px">&nbsp;</div><div id="chr_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr_ChrA_C_glabrata_CBS138">Distribution of variants on chromosome ChrA_C_glabrata_CBS138</h3><div id="chr_ChrA_C_glabrata_CBS138_area" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr_ChrB_C_glabrata_CBS138">Distribution of variants on chromosome ChrB_C_glabrata_CBS138</h3><div id="chr_ChrB_C_glabrata_CBS138_area" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="protein">Position in protein</h3><div id="protein_bar" style="width: 800px; height: 200px">&nbsp;</div></div>
</div></body>
</html>
